src/olympia/amo/middleware.py

```diff
@@ -279,38 +280,70 @@ def process_exception(self, request, exception):
 
 class SetRemoteAddrFromForwardedFor(MiddlewareMixin):
     """
-    Set REMOTE_ADDR from HTTP_X_FORWARDED_FOR if the request came from the CDN.
-
-    If the request came from the CDN, the client IP will always be in
-    HTTP_X_FORWARDED_FOR in second to last position (CDN puts it last, but the
-    load balancer then appends the IP from where it saw the request come from,
-    which is the CDN server that forwarded the request). In addition, the CDN
-    will set HTTP_X_REQUEST_VIA_CDN to a secret value known to us so we can
-    verify the request did originate from the CDN.
-
-    If the request didn't come from the CDN and is a direct origin request, the
-    client IP will be in last position in HTTP_X_FORWARDED_FOR but in this case
-    nginx already sets REMOTE_ADDR so we don't need to use it.
+    Set REMOTE_ADDR from HTTP_X_FORWARDED_FOR if necessary.
+
+    The request flow is either:
+    Client -> CDN -> Load balancer -> WSGI proxy -> addons-server.
+    or
+    Client -> Load balancer -> WSGI proxy -> addons-server.
+
+    Currently:
+    - CDN is CloudFront
+    - Load Balancer is either a classic ELB (AWS) or GKE Ingress (GCP)
+    - WSGI proxy is nginx + uwsgi
+
+    CloudFront is set up to add a X-Request-Via-CDN header set to a secret
+    value known to us so we can verify the request did originate from the CDN.
+
+    Nginx converts X-Request-Via-CDN and X-Forwarded-For to
+    HTTP_X_REQUEST_VIA_CDN and HTTP_X_FORWARDED_FOR, respectively.
+
+    CloudFront always make make origin requests with a X-Forwarded-For header
+    set to "Client IP, CDN IP", so the client IP will be second to last for a
+    CDN request.
+
+    On top of that, on AWS, the classic ELB we're using does not make any
+    alterations to X-Forwarded-For but on GCP, GKE Ingress appends its own IP,
+    so the client IP would move from second to last to third to last in the CDN
+    case
+
+    If the request didn't come from the CDN and is a direct origin request, on
+    AWS the client IP will be in last position in HTTP_X_FORWARDED_FOR but in
+    this case nginx already sets REMOTE_ADDR so we don't need to use it. On GCP
+    though, that won't be true and we'll need to extract it from the second to
+    last position in HTTP_X_FORWARDED_FOR.
```

Comment:
This is super confusing - the request flow section above was clear, so can you write out where what the order of the IPs will be in the same way for the 3(?) scenarios?

Reply:
Reworded in dec7f275bee19d5893628a55c8160dbe00d207b0 let me know if that works for you

Reply:
Yeah, it's clearer.  Though it looks like it's always just the first IP address?  If it's more complicated than that we should probably mention why it's not just the first IP address in the list.

Reply:
That's just in general because the beginning of the contents of the header can't be trusted, it's potentially user input. I'll add a note to that.

Reply:
21bd46f5a6ca4a377b07a68b1a1655e6273a3d9f